commitment_scheme/src/merkle/mod.rs

```diff
@@ -0,0 +1,2014 @@
+) -> Vec<H::Output> {
+    let n_elements = bytes_data.len() / size_of_element;
+    let mut bytes_as_hash = Vec::with_capacity(n_elements);
+
+    for hash_idx in 0..n_elements {
+        let offset = hash_idx * size_of_element;
+        let chunk = &bytes_data[offset..offset + size_of_element];
+        let hash: GenericArray<u8, U32> = GenericArray::clone_from_slice(chunk);
+        bytes_as_hash.push(hash);
+    }
+
+    bytes_as_hash
+}
+
+/// Errors returned by Merkle Tree Verification
+#[derive(Clone, Debug, Eq, PartialEq, Error)]
+#[non_exhaustive]
+pub enum MerkleError {
+    /// returned if error in merkle proof verification
+    #[error("VerificationFail")]
+    VerificationFail,
+    /// returned if query to generate the sibilings is invalid
+    #[error("InvalidQuery")]
+    InvalidQuery,
+}
+
+#[cfg(test)]
+mod tests {
+    use ark_ff::PrimeField;
+    use poseidon::Poseidon3;
+
+    use crate::{
+        merkle::hash::{Blake2s256Hasher, Hasher, Keccak256Hasher},
+        merkle::MerkleTree,
+    };
+
+    use felt::Felt252;
+    use generic_array::typenum::U32;
+    use generic_array::GenericArray;
+    use hex::decode;
+    use rand::Rng;
+
+    use channel::fs_verifier_channel::FSVerifierChannel;
+    use channel::{fs_prover_channel::FSProverChannel, ProverChannel};
+    use randomness::keccak256::PrngKeccak256;
+    use randomness::Prng;
+    use sha3::Sha3_256;
+
+    fn hex_to_generic_array(hex_str: &str) -> GenericArray<u8, U32> {
+        let mut hex_str = String::from(hex_str);
+        let padding_length = 64_i32.saturating_sub(hex_str.len() as i32);
+        if padding_length > 0 {
+            let padding = "0".repeat(padding_length as usize);
+            hex_str.insert_str(0, &padding);
+        }
+        let mut bytes = decode(hex_str).unwrap();
+
+        let padding_length = 32_i32.saturating_sub(bytes.len() as i32);
+        if padding_length > 0 {
+            let mut padding = vec![0u8; padding_length as usize];
+            padding.append(&mut bytes);
+            bytes = padding;
+        }
+        assert_eq!(bytes.len(), 32);
+
+        let mut array = GenericArray::<u8, U32>::default();
+        array.clone_from_slice(&bytes);
+        array
+    }
+
+    fn test_diff_tree_diff_root<F, H>(data0: &[H::Output], data1: &[H::Output])
+    where
+        F: PrimeField,
+        H: Hasher<F, Output = GenericArray<u8, U32>>,
+    {
+        let height = data0.len().ilog2() as usize;
+        let mut tree = MerkleTree::<F, H>::new(data0.len());
+
+        MerkleTree::add_data(&mut tree, data0, 0);
+        let root0 = tree.get_root(0);
+
+        MerkleTree::add_data(&mut tree, data1, 0);
+        let root1 = tree.get_root(height);
+
+        assert!(root0 != root1);
+    }
+
+    #[test]
+    fn test_diff_tree_diff_root_with() {
+        // test with Blake2s256
+        let data0 = [
+            hex_to_generic_array(
+                "837cd6db8f93d5ddfeaaba5509551c69c760c099e97969cfa04047ae65c33c8c",
```

Comment:
Can we reuse these constants across different test cases? (i.e. instantiate them once and reuse them?)

Reply:
Combined the tests which used the same test vectors to avoid duplicate instantiation.